versions/development/SPEC.md

```diff
@@ -2327,9 +2327,17 @@ Then the command would be instantiated as:
 /bin/mycmd /path/to/c.txt
 ```
 
-## Prepending a String to an Optional Parameter
+In general, it's invalid to supply an expression of optional type `T?` for a non-optional input or variable of type `T`, as the latter cannot accept `None`. (The idiom `select_first([expr, default])` coerces `expr : T?` to `T` by substituting `default : T` when `expr` is undefined; if `default` is omitted, then the coercion fails at runtime.) Similarly, an `Array[T]` expression does not satisfy an `Array[T]+` input.
 
-Sometimes, optional parameters need a string prefix.  Consider this task:
+These constraints also apply to function arguments, with exceptions for string concatenation, described below, and the equality/inequality operators `==` and `!=`, which can compare two values of types differing only in their quantifiers, considering `None` equal to itself but no other value.
+
+Quantifiers may apply within compound types; for example, if `a : Array[T?]+`, then `a[0] : T?`. An array type can be both non-empty and optional, `Array[T]+?`, such that the value can be `None` or a non-empty array, but not the empty array. (Reliance on this subtle distinction is discouraged, but it arises from other constructs such as a conditional call with an `Array[T]+` output.)
```

Comment:
I don't think it's valid to discourage using `Array[T]+?`, consider you might have an optional input which may occur multiple times (eg. giving multiple bed/interval files to a variantcaller). It's optional, thus the `?`, and giving an empty array might result in an invalid command.

eg:
There are two valid ways of calling a script: `someScript` or `somScript -i 1.txt -i 2.txt ...`
The command block might looks like this: `someScript ~{true="-i " false="" defined(inputs)}~{sep="-i " inputs}`
- Using `Array[File]` or `Array[File]+` would not allow for the first option.
- Using `Array[File]?` would allow for an empty array which would prduce the invalid command `someScript -i`
- So your only option in to use `Array[File]+?`.

Reply:
Sure, I think its fair to say the parentheses statement is more of my editorial opinion and can be stricken.